emily/client-tests/tests/integration/chainstate.rs

```diff
@@ -0,0 +1,179 @@
+    let configuration = clean_setup().await;
+
+    // Arrange.
+    // --------
+    let mut expected_chainstates: Vec<Chainstate> = (min_height..max_height + 1)
+        .map(|height| new_test_chainstate(height, 0))
+        .collect();
+
+    let expected_chaintip = new_test_chainstate(max_height, 0);
+
+    // Act.
+    // --------
+    let mut created_chainstates =
+        batch_set_chainstates(&configuration, expected_chainstates.clone()).await;
+
+    let mut gotten_chainstates: Vec<Chainstate> = Vec::new();
+    for chainstate in expected_chainstates.iter() {
+        let gotten_chainstate = apis::chainstate_api::get_chainstate_at_height(
+            &configuration,
+            chainstate.stacks_block_height,
+        )
+        .await
+        .expect("Received an error after making a valid get chainstate at height api call.");
+        gotten_chainstates.push(gotten_chainstate);
+    }
+
+    let gotten_chaintip = apis::chainstate_api::get_chain_tip(&configuration)
+        .await
+        .expect("Received an error after making a valid set chaintip api call.");
+
+    // Assert.
+    // --------
+    expected_chainstates.sort_by(arbitrary_chainstate_partial_cmp);
+    created_chainstates.sort_by(arbitrary_chainstate_partial_cmp);
+    gotten_chainstates.sort_by(arbitrary_chainstate_partial_cmp);
+    assert_eq!(expected_chainstates, created_chainstates);
+    assert_eq!(expected_chainstates, gotten_chainstates);
+    assert_eq!(expected_chaintip, gotten_chaintip)
+}
+
+#[cfg_attr(not(feature = "integration-tests"), ignore)]
+#[test_case(1123, 1128, 1133; "standard-reorg")]
+#[test_case(1123, 1133, 1133; "reorg-to-tip-at-same-height")]
+#[test_case(1123, 1111, 1133; "reorg-to-tip-below-any-existing-entry")]
+#[tokio::test]
+async fn create_and_get_chainstate_reorg_happy_path(
+    min_height: i64,
+    reorg_height: i64,
+    max_height: i64,
+) {
+    let configuration = clean_setup().await;
+
+    // Arrange.
+    // --------
+    let original_chainstates: Vec<Chainstate> = (min_height..max_height + 1)
+        .map(|height| new_test_chainstate(height, 0))
+        .collect();
+
+    let expected_post_reorg_chaintip = new_test_chainstate(reorg_height, 1);
+
+    // Act.
+    // --------
+    batch_set_chainstates(&configuration, original_chainstates.clone()).await;
+
+    let created_reorged_chainstate =
+        apis::chainstate_api::set_chainstate(&configuration, expected_post_reorg_chaintip.clone())
+            .await
+            .expect("Received an error after making a valid set chainstate api call.");
+
+    let gotten_post_reorg_chaintip = apis::chainstate_api::get_chain_tip(&configuration)
+        .await
+        .expect("Received an error after making a valid get chaintip api call.");
+
+    // Assert.
+    // --------
+    assert_eq!(expected_post_reorg_chaintip, created_reorged_chainstate);
+    assert_eq!(expected_post_reorg_chaintip, gotten_post_reorg_chaintip);
```

Comment:
At the end of this test we check that the chain tip is at the `reorg_height`, but isn't that incorrect? Why doesn't the chain tip always point to the chain with the max length? I think this implies that a new block with a height that is lower than the current max height is not really a reorg yet.

Reply:
It actually is a reorg! The API follows the event observer, and the event observer will only ever report a burn block if it's now considered the canonical chain tip. The moment there's a conflicting block that's lower than the existing one it should be considered a reorg.